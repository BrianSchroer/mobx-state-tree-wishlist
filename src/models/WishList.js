import { types, getParent, destroy } from 'mobx-state-tree';

export const WishListItem = types
  .model({
    name: types.optional(types.string, ''),
    price: types.optional(types.number, 0),
    image: '' // optional string - default to '' - shorthand for "image: types.optional(types.string, '')"
  }
  ).actions(self => ({
    changeName: newName => self.name = newName,
    changePrice: newPrice => self.price = newPrice,
    changeImage: newImage => self.image = newImage,
    remove: () => getParent(self, 2).remove(self) // 2 levels up: items, then WishList
  }));

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  }
  ).actions(self => ({
    add: item => self.items.push(item),
    remove: item => destroy(item)
  }))
  .views(self => ({
    get totalPrice() {
      return self.items.reduce((sum, item) => sum + item.price, 0);
    }
  }));