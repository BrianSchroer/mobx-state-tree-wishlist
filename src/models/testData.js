export const testWishListItemInput = {
  name: 'Chronicles of Narnia Box Set - C.S. Lewis',
  price: 28.83,
  image: 'https://images-na.ssl-images-amazon.com/images/I/51LmtX5KPAL._SX406_BO1,204,203,200_.jpg'
};

export const testWishListItemInputs =
  [1, 2, 3].map(i => ({ name: `name${i}`, price: i, image: `image${i}` }));

export const testUserInput = {
  id: '1',
  name: 'Brian',
  gender: 'm'
}

export const testUserInputWithWishList = {
  id: '1',
  name: 'Brian',
  gender: 'm',
  wishList: {
    items: testWishListItemInputs
  }
}