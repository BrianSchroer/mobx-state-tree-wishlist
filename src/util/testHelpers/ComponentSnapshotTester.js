import React from 'react';
import renderer from 'react-test-renderer';

/**
 * React component snapshot helper
 */
export default class ComponentSnapshotTester {

  /**
   * Create new ComponentSnapshotTester instance
   * @param {*} component - React component
   */
  constructor(component) {
    this.component = component.type;
    this.props = component.props;
    this.adjustProps = (props, propOverrides) => Object.assign({}, props, propOverrides);
  }

  /**
   * Specify callback function to be called by the test function to adjust component
   * props with propOverrides.
   * (Defaults to adjusting top-level props.)
   */
  withPropAdjustor = (adjustor) => {
    this.adjustProps = adjustor;
    return this;
  };

  /**
   * Adjust props that were passed to the constructor with propOverrides, then create
   * snapshot and assert that it matches previously accepted screenshot.
   */
  test = (propOverrides) => {
    const Component = this.component;
    const props = this.adjustProps(this.props, propOverrides);

    const newSnapshot = renderer.create(<Component {...props} />).toJSON();

    expect(newSnapshot).toMatchSnapshot();
  }

  /**
   * Create snapshot and assert that it matches previously accepted screenshot. 
   * @param {*} component - React component1 
   */
  static test(component) {
    new ComponentSnapshotTester(component).test();
  }
}