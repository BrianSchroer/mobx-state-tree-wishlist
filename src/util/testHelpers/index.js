export { default as mobxSnapshotHelper } from './mobxSnapshotHelper/snapshotHelper';
export { default as ComponentSnapshotHelper } from './ComponentSnapshotHelper/SnapshotHelper';