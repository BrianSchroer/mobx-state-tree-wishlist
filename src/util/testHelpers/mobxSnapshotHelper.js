import { getSnapshot } from 'mobx-state-tree';

export default class {
  static createSnapshot(target) {
    return getSnapshot(target);
  }

  static assertMatch(target) {
    expect(this.createSnapshot(target)).toMatchSnapshot();
  }
}