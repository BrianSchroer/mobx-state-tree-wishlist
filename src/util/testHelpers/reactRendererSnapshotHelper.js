import renderer from 'react-test-renderer';

/**
 * react-test-renderer snapshot helper methods
 */
export default class {

  /**
   * Create component from JSX
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {object} component object
   */
  static createComponent(componentJsx) {
    return renderer.create(componentJsx);
  }

  /**
   * Create component from JSX and serialize as JSON "snapshot" string.
   * @param {JSX} componentJsx (e.g. '<Button />')
   * @returns {string} JSON string.
   */
  static createSnapshot(componentJsx) {
    return this.createComponent(componentJsx).toJSON();
  }

  /**
   * Create component from JSX, serialize as JSON string, and compare to accepted snapshot.
   * If a snapshot doesn't already exist, one is saved. If a previous snapshot exists
   * and the new snapshot doesn't match, you'll be asked if that's an error or if you want
   * to accept the new snapshot.
   * @param {JSX} componentJsx (e.g. '<Button />')
   */
  static assertMatch(componentJsx) {
    expect(this.createSnapshot(componentJsx)).toMatchSnapshot();
  }
}