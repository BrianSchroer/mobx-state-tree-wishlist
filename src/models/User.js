import { types } from 'mobx-state-tree';
import { WishList } from './WishList';

export const User = types.model({
  id: types.string,
  name: types.string,
  gender: types.enumeration('gender', ['m', 'f', 'x']),
  wishList: types.optional(WishList, {})
});