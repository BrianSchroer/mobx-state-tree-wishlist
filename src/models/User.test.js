import { reaction } from 'mobx';
import { User } from './User';
import { WishList } from './WishList';
import { testUserInput, testUserInputWithWishList, testWishListItemInputs } from './testData';
import { mobxSnapshotHelper as snapshotHelper } from '../util/testHelpers';

function testUserInputWith(overrides) {
  return Object.assign({}, testUserInput, overrides);
}

describe('User', () => {
  describe('.create()', () => {
    it('should return expected item without wishlist', () => {
      snapshotHelper.test(User.create(testUserInput));
    });

    it('should return expected item with wishlist', () => {
      snapshotHelper.test(User.create(testUserInputWithWishList));
    });

    ['m', 'f', 'x'].forEach(gender => {
      it(`should accept gender "${gender}"`, () => {
        snapshotHelper.test(User.create(testUserInputWith({ gender })));
      });
    });

    it('should not accept unknown genders', () => {
      try {
        snapshotHelper.test(User.create(testUserInputWith({ gender: 'q' })));
      } catch (error) {
        expect(error.toString()).toEqual(expect.stringContaining(
          '"/gender" value `"q"` is not assignable to type: `gender`'));
      }
    })
  });
});
