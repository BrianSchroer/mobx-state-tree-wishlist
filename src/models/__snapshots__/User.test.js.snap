// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User .create() should accept gender "f" 1`] = `
Object {
  "gender": "f",
  "id": "1",
  "name": "Brian",
  "wishList": Object {
    "items": Array [],
  },
}
`;

exports[`User .create() should accept gender "m" 1`] = `
Object {
  "gender": "m",
  "id": "1",
  "name": "Brian",
  "wishList": Object {
    "items": Array [],
  },
}
`;

exports[`User .create() should accept gender "x" 1`] = `
Object {
  "gender": "x",
  "id": "1",
  "name": "Brian",
  "wishList": Object {
    "items": Array [],
  },
}
`;

exports[`User .create() should return expected item with wishlist 1`] = `
Object {
  "gender": "m",
  "id": "1",
  "name": "Brian",
  "wishList": Object {
    "items": Array [
      Object {
        "image": "image1",
        "name": "name1",
        "price": 1,
      },
      Object {
        "image": "image2",
        "name": "name2",
        "price": 2,
      },
      Object {
        "image": "image3",
        "name": "name3",
        "price": 3,
      },
    ],
  },
}
`;

exports[`User .create() should return expected item without wishlist 1`] = `
Object {
  "gender": "m",
  "id": "1",
  "name": "Brian",
  "wishList": Object {
    "items": Array [],
  },
}
`;
